Override module chrome with a plugin

jrseliga opened the pull request:
#### Purpose

Template providers often ship the same module chromes with every template. It is much more convenient to manage one file and ship a plugin or a plugin + library that could register these module chromes.

Unfortunately a popular template framework provider already implements this concept by overriding
the core file /libraries/cms/module/helper.php with their framework initialization plugin for the purposes of adding their own triggerEvent.
#### How it works

A plugin that executes on the **onGetModuleChromes** must return an array with keys **element** (friendly name for their list of module chromes) and **path** (path to file containing module chrome definitions)
#### Testing
1. Install Joomla (sample data not necessary)
2. [Upload plugin folders to /plugins/system](https://github.com/jrseliga/module-chrome-plugins) 
3. Discover / Enable plugins **modulechromesone** and **modulechromestwo**
4. Navigate to front-end notice the "Login Form" module
   ![default-install](https://cloud.githubusercontent.com/assets/3277067/8049245/bc724da0-0e2a-11e5-966c-2700b522c0d0.png)
5. Module Manager
   - Login Form
     1. Advanced tab
     2. Module Style
        - Scroll to bottom, will see two new grouped lists **Plugin One** and **Plugin Two**
          ![module-style](https://cloud.githubusercontent.com/assets/3277067/8049249/c522ed10-0e2a-11e5-8b4f-0a4e052cc833.png)
   - Choose any from either of these groups
   - Refresh front-end, the name of the module chrome function used will be displayed in place of the log in form.![one](https://cloud.githubusercontent.com/assets/3277067/8049264/f90ebaf0-0e2a-11e5-915a-0175693a933f.png)
   - Repeat last two steps to see different chromes rendered. ![four](https://cloud.githubusercontent.com/assets/3277067/8049255/dcae7daa-0e2a-11e5-944a-d0ed79fc135a.png)

jrseliga committed:
Override module layout with a plugin

Bakual commented:
First you need to fix some codestyle issue. Between the `if` and the opening `(` there has to be a space. So instead of `if(...)` make it `if (...)`.

To get an idea how this should work and to be able to test it, can you add an example plugin which would contain a few overrides?

Also, please change the order. The template override always should have the last word. With your proposal the plugin would override the template override which would be wrong.

zero-24 committed:
fix the undefined $app; Some code simplefication; Codestyle; The template override always should have the last word.

zero-24 committed:
fix the comments

zero-24 commented:
@jrseliga @Bakual here is a PR that should fix the described issues and also with some simplications: https://github.com/jrseliga/joomla-cms/pull/1

jrseliga commented:
@Bakual @zero-24 forgetting $app was a poor copy-paste from my development site.

Looks like @zero-24 fixed the issues, I'll merge that PR.

@Bakual, I will provide a plugin for testing, first I'd like to update /libraries/cms/form/field/chromestyle.php so that the module's configuration can see these chromes as options in "Module Style".

zero-24 committed:
CS

jrseliga committed:
Merge pull request #1 from zero-24/patch-1

define $app, code simplification, code style, give template override highest priority

jrseliga committed:
Moved system and active template chrome paths into an array. Allow plugins to include chrome paths. Iterate chrome path array, if file exists register it as true.

Implemented array holding module layout paths. Appends new array items in appropriate priority order, when iterating over array, path returned follows template, plugin, then module layouts.

jrseliga committed:
More suitable event names

jrseliga commented:
Originally this was also going to include the ability to override a layout for modules via a plugin, however testing revealed the dependency this has on plugin priority. Injecting a layout in this way is not conducive, or predictable for developers or site administrators. Will remove this event.

Despite the branch name this will only focus on module chromes as the general problem I'd like to solve is template providers needing to change manage the same /templates/template/html/modules.php across all of their templates. If you reuse the same chrome, ship your template with a library and a plugin (or just a plugin) that registers your a path to your chrome file(s). If it matches the Joomla! chrome function naming convention it will be registered.

jrseliga committed:
Code style & removed event for adding module layout via plugin.

jrseliga commented:
Layout override via plugin removed for reason above. Will supply a plugin for testing.

jrseliga committed:
No longer need to declare $app, waste of memory.

Bakual commented on libraries/cms/module/helper.php:
```diff
@@ -198,17 +198,23 @@ public static function renderModule($module, $attribs = array())
 
-		if (!isset($chrome[$chromePath]))
+		// Load chromes from plugins
+		$chromePaths = array_merge($chromePaths, $app->triggerEvent('onGetModuleChromePaths'));
```
I'd first check if the plugin returns indeed a valid array of paths before merging it.
Also you need to think about what happens when there is more than one plugin returning a path.
Maybe you need to pass the plugins an array which will be filled with custom paths.

jrseliga committed:
The same event 'onGetModuleChromes' should be used in both files instead of separate events per file. Made changes for the sake of consistency.

jrseliga commented:
@Bakual significant changes were made, the original comment will be rewritten with instructions and more detailed information.

I have two plugins ready for testing to prove it can handle multiple, and have successfully tested functionality.
Should I upload them as a separate repository or is there a way to include them in this branch without the PR noticing them?

Bakual commented:
I would just put the plugins somewhere available for download. Like Dropbox or your own website. The put the link into the PR here.

phproberto commented on libraries/cms/module/helper.php:
```diff
@@ -278,35 +302,30 @@ public static function renderModule($module, $attribs = array())
-		else
-		{
-			return $dPath;
+			if (file_exists($layoutPath))
```
You should use `JPath::find()` here

phproberto commented:
While I like the idea if at some point `JLayout` is going to be used for this it's probably a good idea to wait for it. This is a new feature so v3.5 is the minimum release for it anyway.

I see a lof of file_exists() checks that could be avoided with `JLayout` or at least with `JPath`

nikosdion commented:
I love the concept and I agree with Roberto. Using JLayout would be the future-proof way to implement such a fundamental feature, especially considering the key role JLayout is poised to have in Joomla! 4 (at least according to the _informal_ discussion we had with half of the PLT and some Joomla! contributors during the Make It Happen session at JaB 15).

Roberto can you help Justin implement that? You're the de facto JLayout expert, the rest of us are mere apprentices :)

jrseliga commented:
@phproberto would like to see this available in 3.5 and would be happy to help implement the concept with `JLayout`.

hardiktailored commented:
I have tested this item :white_check_mark: successfully on d90e033fc3cee41c4929f7b84eaff9f919e66eb2<br /><br />Followed the testing steps and working as explained.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/7099">issues.joomla.org/joomla-cms/7099</a>.</sub>

roland-d commented:
@jrseliga Could you update this PR to use JLayout? Thank you.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/7099">issues.joomla.org/joomla-cms/7099</a>.</sub>

BhaveshTailored commented:
I have tested this item :white_check_mark: successfully on d90e033fc3cee41c4929f7b84eaff9f919e66eb2<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/7099">issues.joomla.org/joomla-cms/7099</a>.</sub>

brianteeman commented:
I am closing this at this time as it appears to have been abandoned by @jrseliga 
It can always be re-opened if updated as per @roland-d comments above
